parser/openfoodfacts_taxonomy_parser/parser.py

```diff
@@ -0,0 +1,336 @@
+        if lc in self.stopwords:
+            words_to_remove = self.stopwords[lc]
+            new_words = []
+            for word in words.split("-"):
+                if word not in words_to_remove:
+                    new_words.append(word)
+            return ("-").join(new_words)
+        else:
+            return words
+
+    def add_line(self, line):
+        """to get a normalized string but keeping the language code "lc:" , used for id and parent tag"""
+        lc, line = line.split(":", 1)
+        new_line = lc + ":"
+        new_line += self.remove_stopwords(lc, self.normalizing(line, lc))
+        return new_line
+
+    def get_lc_value(self, line):
+        """to get the language code "lc" and a list of normalized values"""
+        lc, line = line.split(":", 1)
+        new_line = []
+        for word in line.split(","):
+            new_line.append(self.remove_stopwords(lc, self.normalizing(word, lc)))
+        return lc, new_line
+
+    def new_node_data(self):
+        """To create an empty dictionary that will be used to create node"""
+        data = {
+            "id": "",
+            "main_language": "",
+            "preceding_lines": [],
+            "parent_tag": [],
+            "src_position": 0,
+        }
+        return data
+
+    def set_data_id(self, data, id, line_number):
+        if not data["id"]:
+            data["id"] = id
+        else:
+            raise DuplicateIDError(line_number)
+        return data
+
+    def header_harvest(self, filename):
+        """to harvest the header (comment with #), it has its own function because some header has multiple blocks"""
+        h = 0
+        header = []
+        for _, line in self.file_iter(filename):
+            if not (line) or line[0] == "#":
+                header.append(line)
+            else:
+                break
+            h += 1
+
+        # we don't want to eat the comments of the next block and it remove the last separating line
+        for i in range(len(header)):
+            if header.pop():
+                h -= 1
+            else:
+                break
+
+        return header, h
+
+    def entry_end(self, line, data):
+        """Return True if the block ended"""
+        if "stopwords" in line or "synonyms" in line or not line:
+            # can be the end of an block or just 2 line separators,
+            # file_iter() always end with ''
+            if data["id"]:  # to be sure that it's an end
+                return True
+        return False
+
+    def remove_separating_line(self, data):
+        if data["preceding_lines"]:
+            if "synonyms" in data["id"]:
+                if "stopwords" in self.is_before:
+                    data["preceding_lines"].pop(0)
+            elif "stopwords" in data["id"]:
+                if "synonyms" in self.is_before:
+                    data["preceding_lines"].pop(0)
+            else:
+                data["preceding_lines"].pop(0)
```

Comment:
Oh, this is because stopwords and synonyms often have no separating line between them, so I added the blank line in preceding_lines (when there is one) and I don't want it to be removed.
I'll add some comments